eventstore/query_json.go

```diff
@@ -0,0 +1,232 @@
+		if err != nil {
+			return nil, fmt.Errorf("error when matching entry with query: %v", err)
+		}
+		if ok {
+			values = append(values, marshaledValue{value: val, rawJson: res.Value})
+		}
+	}
+
+	if q.Sort.FieldPath != "" {
+		var wrongField, cantCompare bool
+		sort.Slice(values, func(i, j int) bool {
+			fieldI, err := traverseFieldPathMap(values[i].value, q.Sort.FieldPath)
+			if err != nil {
+				wrongField = true
+				return false
+			}
+			fieldJ, err := traverseFieldPathMap(values[j].value, q.Sort.FieldPath)
+			if err != nil {
+				wrongField = true
+				return false
+			}
+			res, err := compare(fieldI.Interface(), fieldJ.Interface())
+			if err != nil {
+				cantCompare = true
+				return false
+			}
+			if q.Sort.Desc {
+				res *= -1
+			}
+			return res < 0
+		})
+		if wrongField {
+			return nil, ErrInvalidSortingField
+		}
+		if cantCompare {
+			panic("can't compare while sorting")
+		}
+	}
+
+	res := make([]string, len(values))
+	for i := range values {
+		res[i] = string(values[i].rawJson)
+	}
+
+	return res, nil
+}
+
+func (q *JSONQuery) matchJSON(v map[string]interface{}) (bool, error) {
+	if q == nil {
+		panic("query can't be nil")
+	}
+
+	andOk := true
+	for _, c := range q.Ands {
+		fieldRes, err := traverseFieldPathMap(v, c.FieldPath)
+		if err != nil {
+			return false, err
+		}
+		ok, err := c.match(fieldRes)
+		if err != nil {
+			return false, err
+		}
+		andOk = andOk && ok
+		if !andOk {
+			break
+		}
+	}
+	if andOk {
+		return true, nil
+	}
+
+	for _, q := range q.Ors {
+		ok, err := q.matchJSON(v)
+		if err != nil {
+			return false, err
+		}
+		if ok {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func compareJsonValue(value interface{}, critVal JSONValue) (int, error) {
+	if critVal.String != nil {
+		s, ok := value.(string)
+		if !ok {
+			return 0, &errTypeMismatch{value, critVal}
+		}
+		return strings.Compare(s, *critVal.String), nil
+	}
+	if critVal.Bool != nil {
+		b, ok := value.(bool)
+		if !ok {
+			return 0, &errTypeMismatch{value, critVal}
+		}
+		if *critVal.Bool == b {
+			return 0, nil
+		}
+		return -1, nil
+	}
+	if critVal.Float != nil {
+		f, ok := value.(float64)
+		if !ok {
+			return 0, &errTypeMismatch{value, critVal}
+		}
+		if f == *critVal.Float {
+			return 0, nil
+		}
+		if f < *critVal.Float {
+			return -1, nil
+		}
+		return 1, nil
+	}
+	log.Fatalf("no underlying value for json criterion was provided")
+	return 0, nil
+}
+
+func (c *JSONCriterion) match(value reflect.Value) (bool, error) {
+	valueInterface := value.Interface()
+	result, err := compareJsonValue(valueInterface, c.Value)
+	if err != nil {
+		return false, err
+	}
+	switch c.Operation {
+	case Eq:
+		return result == 0, nil
+	case Ne:
+		return result != 0, nil
+	case Gt:
+		return result > 0, nil
+	case Lt:
+		return result < 0, nil
+	case Le:
+		return result < 0 || result == 0, nil
+	case Ge:
+		return result > 0 || result == 0, nil
+	default:
+		panic("invalid operation")
```

Comment:
Also, notice that JSONMode queries don't have arbitrary `func` operations as in Go Type mode.
In Go Type [you can provide a custom func to sort your data](https://github.com/textileio/go-textile-threads/blob/master/eventstore/query_test.go#L74). This could be done also in JSONMode, but since now will be used by gRPC API doesn't make sense. (If #131 is done, we can add it)